src/Simulation/Simulators/ToffoliSimulator/Processor.cs

```diff
@@ -0,0 +1,429 @@
+
+namespace Microsoft.Quantum.Simulation.Simulators
+{
+
+    // ToffoliSimulator is implemented based on `QuantumProcessorDispatcher`,
+    // which is constructed using a specialization `QuantumProcessorBase`.  The
+    // specialization overrides methods to specify actions on intrinsic
+    // operations in the Q# code.
+    class ToffoliSimulatorProcessor : QuantumProcessorBase {
+        // This property controls whether to throw an exception when a qubit is
+        // released that is not in the zero state.
+        public bool ThrowOnReleasingQubitsNotInZeroState { get; set; } = true;
+
+        // The simulation values are stored in the bits of a BitArray, which
+        // can use integers to store several bits at once.  We initialize the
+        // bit array to contain up to 64 bits and will dynamically grow it,
+        // whenever this capacity is exceeded.
+        public BitArray SimulationValues { get; private set; }
+
+        public ToffoliDumpFormat DumpFormat { get; set; } = ToffoliDumpFormat.Automatic;
+
+        // A property that must be assigned from the dispatcher to return the allocated
+        // qubit ids from the qubit manager
+        public Func<IEnumerable<long>>? GetAllocatedIds { get; set; }
+
+        // A property that must be assigned from the dispatcher to print a Q# message
+        public Action<String>? Message { get; set; }
+
+        private Dictionary<Qubit, Stack<bool>> borrowedQubitStates { get; } = new Dictionary<Qubit, Stack<bool>>();
+        private const double Tolerance = 1.0E-10;
+
+        public ToffoliSimulatorProcessor(uint qubitCount, bool throwOnReleasingQubitsNotInZeroState) {
+            SimulationValues = new BitArray((int)qubitCount);
+            ThrowOnReleasingQubitsNotInZeroState = throwOnReleasingQubitsNotInZeroState;
+        }
+
+        // Three helper functions manipulate the bits in the bit array given a
+        // qubit as input. GetValue returns the simulation value at the
+        // corresponding index.
+        private bool GetValue(Qubit qubit) =>
+            SimulationValues[qubit.Id];
+
+        // SetValue sets the simulation value at the corresponding index.
+        private void SetValue(Qubit qubit, bool value) {
+            SimulationValues[qubit.Id] = value;
+        }
+
+        // InvertValue inverts the simulation value at the corresponding index.
+        private void InvertValue(Qubit qubit) {
+            SimulationValues[qubit.Id] = !SimulationValues[qubit.Id];
+        }
+
+        // When allocating a qubit whose index equals or exceeds the current
+        // number of simulation values stored in the bit array, the bit array is
+        // resized to twice its current size.
+        public override void OnAllocateQubits(IQArray<Qubit> qubits) {
+            if (qubits.Count == 0) return;
+
+            var maxId = qubits.Max(q => q.Id);
+            while (maxId >= SimulationValues.Length) {
+                SimulationValues.Length *= 2;
```

Comment:
>SimulationValues.Length *= 2; [](start = 16, length = 29)

What if the doubled size is still less than maxId?

Reply:
It'll double until it's larger than `maxId` using the condition in the `while` loop.

Reply:
Duh! Of course :)